File: modules/reliquary/legacy/python/src/reliquary/store.py
"""Local staging store: ingest, list, verify, extract."""

from __future__ import annotations

import shutil
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from reliquary.config import Config
from reliquary import archive, hashing, manifest, parity


class StoreError(RuntimeError):
    pass


def _date_prefix() -> str:
    return datetime.now(timezone.utc).strftime("%Y%m%d")


def make_block_id(payload: Path) -> str:
    digest = hashing.hash_file(payload)["sha256"][:16]
    return f"{_date_prefix()}-{digest}"


class Store:
    def __init__(self, cfg: Config):
        self.cfg = cfg
        cfg.ensure_dirs()
        if not cfg.catalog_path.exists():
            manifest.write_json(cfg.catalog_path, manifest.empty_catalog())

    def catalog(self) -> dict[str, Any]:
        return manifest.read_json(self.cfg.catalog_path)

    def _save_catalog(self, cat: dict[str, Any]) -> None:
        cat["updated"] = manifest.utcnow()
        manifest.write_json(self.cfg.catalog_path, cat)

    def block_dir(self, block_id: str) -> Path:
        return self.cfg.blocks / block_id

    def load_manifest(self, block_id: str) -> dict[str, Any]:
        path = self.block_dir(block_id) / "manifest.json"
        if not path.exists():
            raise StoreError(f"unknown block {block_id}")
        return manifest.read_json(path)

    def list_blocks(self) -> list[dict[str, Any]]:
        rows = []
        for child in sorted(self.cfg.blocks.iterdir()) if self.cfg.blocks.exists() else []:
            man = child / "manifest.json"
            if man.exists():
                rows.append(manifest.read_json(man))
        return rows

    def ingest(
        self,
        source: Path,
        *,
        profile: str = "cd",
        notes: str = "",
        force: bool = False,
    ) -> dict[str, Any]:
        source = source.expanduser().resolve()
        if not source.exists():
            raise StoreError(f"source does not exist: {source}")

        raw = archive.tree_bytes(source)
        if profile == "cd" and raw > self.cfg.cd_payload_bytes * 4:
            # Uncompressed source is huge; compressed size is only known
            # after packing. We still pack, then refuse to keep if the
            # finished payload will not fit a CD with PAR2.
            pass

        work = self.cfg.work_root / "ingest"
        if work.exists():
            shutil.rmtree(work)
        work.mkdir(parents=True)

        payload = work / "payload.tar.zst"
        packed = archive.pack_tree(source, payload, zstd_level=self.cfg.zstd_level)
        digests = hashing.hash_file(payload)
        packed.update(digests)

        if profile == "cd":
            # payload + ~20% PAR2 + ~2 MiB manifests must fit the disc.
            budget = int(self.cfg.cd_capacity_bytes * 0.92)
            expected = int(packed["bytes"] * (1 + self.cfg.par2_redundancy / 100) + 2 * 1024 * 1024)
            if expected > budget and not force:
                shutil.rmtree(work)
                raise StoreError(
                    f"packed payload {packed['bytes']} B plus {self.cfg.par2_redundancy}% PAR2 "
                    f"will not fit an 80-minute CD-R (~{budget} B usable). "
                    "Split the source, use --profile usb, or pass --force."
                )

        block_id = make_block_id(payload)
        dest = self.block_dir(block_id)
        if dest.exists():
            shutil.rmtree(work)
            return self.load_manifest(block_id)

        dest.mkdir(parents=True)
        shutil.move(str(payload), dest / "payload.tar.zst")
        payload = dest / "payload.tar.zst"

        par = parity.create_par2(
            payload,
            redundancy=self.cfg.par2_redundancy,
            volumes=self.cfg.par2_volumes,
        )
        hashing.write_sum_files(
            dest,
            [payload, *sorted(dest.glob("*.par2"))],
        )
        # Re-write sums after PAR2 so the index itself is covered? PAR2
        # files are listed above. Manifest is written last and hashed into
        # a sidecar so the document can be edited when copies are recorded.
        man = manifest.new_block_manifest(
            block_id=block_id,
            origin=str(source),
            payload=packed,
            par2=par,
            profile=profile,
            notes=notes,
        )
        manifest.write_json(dest / "manifest.json", man)
        hashing.write_sum_files(
            dest,
            [p for p in dest.iterdir() if p.is_file() and p.name not in {"SHA256SUMS", "SHA512SUMS"}],
        )

        cat = self.catalog()
        cat["blocks"] = [b for b in cat.get("blocks", []) if b.get("id") != block_id]
        cat["blocks"].append(
            {
                "id": block_id,
                "created": man["created"],
                "origin": man["origin"],
                "profile": profile,
                "payload_bytes": packed["bytes"],
                "sha256": packed["sha256"],
            }
        )
        self._save_catalog(cat)
        shutil.rmtree(work, ignore_errors=True)
        return man

    def verify(self, block_id: str, *, repair: bool = False) -> dict[str, Any]:
        dest = self.block_dir(block_id)
        if not dest.exists():
            raise StoreError(f"unknown block {block_id}")
        problems = hashing.verify_sum_file(dest, "SHA256SUMS")
        problems += hashing.verify_sum_file(dest, "SHA512SUMS")
        index = dest / "payload.tar.zst.par2"
        par = {"ok": False, "output": "missing PAR2 index"}
        if index.exists():
            par = parity.verify_par2(index)
            if not par["ok"] and repair:
                par = parity.repair_par2(index)
                if par["ok"]:
                    par = parity.verify_par2(index)
        ok = not problems and par.get("ok")
        return {"id": block_id, "ok": ok, "checksum_problems": problems, "par2": par}

    def extract(self, block_id: str, dest: Path, *, verify_first: bool = True) -> Path:
        if verify_first:
            result = self.verify(block_id)
            if not result["ok"]:
                raise StoreError(f"block {block_id} failed verification: {result}")
        dest = dest.expanduser().resolve()
        dest.mkdir(parents=True, exist_ok=True)
        archive.extract_payload(self.block_dir(block_id) / "payload.tar.zst", dest)
        return dest

    def copy_block(self, block_id: str, dest_dir: Path, medium: str) -> Path:
        src = self.block_dir(block_id)
        if not src.exists():
            raise StoreError(f"unknown block {block_id}")
        dest_dir = dest_dir.expanduser().resolve()
        dest_dir.mkdir(parents=True, exist_ok=True)
        target = dest_dir / block_id
        if target.exists():
            shutil.rmtree(target)
        shutil.copytree(src, target)
        man = manifest.read_json(target / "manifest.json")
        manifest.record_copy(man, medium, str(target))
        manifest.write_json(target / "manifest.json", man)
        # Also stamp the local original.
        local = manifest.read_json(src / "manifest.json")
        manifest.record_copy(local, medium, str(target))
        manifest.write_json(src / "manifest.json", local)
        return target

    def forget(self, block_id: str, *, yes: bool = False) -> None:
        if not yes:
            raise StoreError("refusing to delete without yes=True")
        dest = self.block_dir(block_id)
        if dest.exists():
            shutil.rmtree(dest)
        cat = self.catalog()
        cat["blocks"] = [b for b in cat.get("blocks", []) if b.get("id") != block_id]
        self._save_catalog(cat)

File: modules/reliquary/src/store.rs
use std::path::{Path, PathBuf};

use serde_json::Value;

use crate::archive;
use crate::config::Config;
use crate::error::{Error, Result};
use crate::hashing;
use crate::manifest;
use crate::parity;

pub struct Store {
    pub cfg: Config,
}

impl Store {
    pub fn new(cfg: Config) -> Result<Self> {
        cfg.ensure_dirs();
        if !cfg.catalog_path().exists() {
            manifest::write_json(&cfg.catalog_path(), &manifest::empty_catalog())?;
        }
        Ok(Self { cfg })
    }

    pub fn catalog(&self) -> Result<Value> {
        manifest::read_json(&self.cfg.catalog_path())
    }

    fn save_catalog(&self, mut cat: Value) -> Result<()> {
        cat["updated"] = Value::String(manifest::utcnow());
        manifest::write_json(&self.cfg.catalog_path(), &cat)
    }

    pub fn block_dir(&self, block_id: &str) -> PathBuf {
        self.cfg.blocks().join(block_id)
    }

    pub fn load_manifest(&self, block_id: &str) -> Result<Value> {
        validate_block_id(block_id)?;
        let path = self.block_dir(block_id).join("manifest.json");
        if !path.exists() {
            return Err(Error::UnknownBlock(block_id.into()));
        }
        manifest::read_json(&path)
    }

    pub fn list_blocks(&self) -> Result<Vec<Value>> {
        let mut rows = Vec::new();
        let blocks = self.cfg.blocks();
        if !blocks.exists() {
            return Ok(rows);
        }
        let mut names: Vec<_> = std::fs::read_dir(&blocks)?
            .flatten()
            .map(|e| e.path())
            .collect();
        names.sort();
        for child in names {
            let man = child.join("manifest.json");
            if man.exists() {
                rows.push(manifest::read_json(&man)?);
            }
        }
        Ok(rows)
    }

    pub fn ingest(&self, source: &Path, profile: &str, notes: &str, force: bool) -> Result<Value> {
        let source = if source.exists() {
            source.canonicalize()?
        } else {
            return Err(Error::MissingSource(source.to_path_buf()));
        };

        let work = self.cfg.work_root.join("ingest");
        crate::util::remove_dir_if_exists(&work)?;
        std::fs::create_dir_all(&work)?;

        let payload_tmp = work.join("payload.tar.zst");
        let packed = archive::pack_tree(&source, &payload_tmp, self.cfg.zstd_level)?;

        if profile == "cd" {
            let budget = (self.cfg.cd_capacity_bytes as f64 * 0.92) as u64;
            let expected = packed.bytes
                + packed.bytes * u64::from(self.cfg.par2_redundancy) / 100
                + 2 * 1024 * 1024;
            if expected > budget && !force {
                let _ = std::fs::remove_dir_all(&work);
                return Err(Error::store(format!(
                    "packed payload {} B plus {}% PAR2 will not fit an 80-minute CD-R (~{} B usable). Split the source, use --profile usb, or pass --force.",
                    packed.bytes, self.cfg.par2_redundancy, budget
                )));
            }
        }

        let block_id = format!(
            "{}-{}",
            chrono::Utc::now().format("%Y%m%d"),
            &packed.sha256[..16]
        );
        let dest = self.block_dir(&block_id);
        if dest.exists() {
            let _ = std::fs::remove_dir_all(&work);
            return self.load_manifest(&block_id);
        }
        std::fs::create_dir_all(&dest)?;
        let payload = dest.join("payload.tar.zst");
        std::fs::rename(&payload_tmp, &payload)?;

        let par = parity::create_par2(&payload, self.cfg.par2_redundancy, self.cfg.par2_volumes)?;

        // Hash only immutable payload + PAR2. Manifest and copies.json are
        // allowed to gain provenance later; including them in SHA256SUMS made
        // every USB push fail verification.
        write_payload_sums(&dest, &payload)?;

        let payload_json = serde_json::to_value(&packed)?;
        let par_json = serde_json::to_value(&par)?;
        let man = manifest::new_block_manifest(
            &block_id,
            &source.to_string_lossy(),
            payload_json,
            par_json,
            profile,
            notes,
        );
        manifest::write_json(&dest.join("manifest.json"), &man)?;
        manifest::write_json(&dest.join("copies.json"), &serde_json::json!([]))?;

        let mut cat = self.catalog()?;
        let blocks = cat
            .get_mut("blocks")
            .and_then(|v| v.as_array_mut())
            .cloned()
            .unwrap_or_default();
        let blocks: Vec<Value> = blocks
            .into_iter()
            .filter(|b| b.get("id").and_then(|v| v.as_str()) != Some(&block_id))
            .collect();
        cat["blocks"] = Value::Array(blocks);
        if let Some(arr) = cat.get_mut("blocks").and_then(|v| v.as_array_mut()) {
            arr.push(serde_json::json!({
                "id": block_id,
                "created": man.get("created"),
                "origin": man.get("origin"),
                "profile": profile,
                "payload_bytes": packed.bytes,
                "sha256": packed.sha256,
            }));
        }
        self.save_catalog(cat)?;
        let _ = std::fs::remove_dir_all(&work);
        Ok(man)
    }

    pub fn verify(&self, block_id: &str, repair: bool) -> Result<Value> {
        validate_block_id(block_id)?;
        let dest = self.block_dir(block_id);
        if !dest.exists() {
            return Err(Error::UnknownBlock(block_id.into()));
        }
        let mut problems = hashing::verify_sum_file(&dest, "SHA256SUMS")?;
        problems.extend(hashing::verify_sum_file(&dest, "SHA512SUMS")?);
        let index = dest.join("payload.tar.zst.par2");
        let mut par = serde_json::json!({"ok": false, "output": "missing PAR2 index"});
        if index.exists() {
            let mut result = parity::verify_par2(&index)?;
            if !result.ok && repair {
                result = parity::repair_par2(&index)?;
                if result.ok {
                    result = parity::verify_par2(&index)?;
                }
            }
            par = serde_json::to_value(result)?;
        }
        let ok = problems.is_empty() && par.get("ok").and_then(|v| v.as_bool()).unwrap_or(false);
        Ok(serde_json::json!({
            "id": block_id,
            "ok": ok,
            "checksum_problems": problems,
            "par2": par,
        }))
    }

    pub fn extract(&self, block_id: &str, dest: &Path, verify_first: bool) -> Result<PathBuf> {
        validate_block_id(block_id)?;
        if verify_first {
            let result = self.verify(block_id, false)?;
            if result.get("ok") != Some(&Value::Bool(true)) {
                return Err(Error::store(format!(
                    "block {block_id} failed verification: {result}"
                )));
            }
        }
        // Refuse a destination that already holds anything. extract_payload's
        // tar-slip check only sees one archive at a time: a first extract can
        // leave a symlink member behind (its own path has no `..`, so it
        // passes), and a second extract into the same dest writes THROUGH that
        // symlink to somewhere outside dest, because tar follows a symlink
        // already sitting at the target. A fresh directory per extract is the
        // only way to close that across two archives.
        if dest.exists() && std::fs::read_dir(dest)?.next().is_some() {
            return Err(Error::store(format!(
                "destination {} is not empty; extract into a fresh, empty directory \
                 (a second extract into a populated directory can be made to follow \
                 a symlink planted by the first)",
                dest.display()
            )));
        }
        std::fs::create_dir_all(dest)?;
        archive::extract_payload(&self.block_dir(block_id).join("payload.tar.zst"), dest)?;
        Ok(dest.to_path_buf())
    }

    pub fn copy_block(&self, block_id: &str, dest_dir: &Path, medium: &str) -> Result<PathBuf> {
        validate_block_id(block_id)?;
        let src = self.block_dir(block_id);
        if !src.exists() {
            return Err(Error::UnknownBlock(block_id.into()));
        }
        std::fs::create_dir_all(dest_dir)?;
        let target = dest_dir.join(block_id);
        crate::util::remove_dir_if_exists(&target)?;
        crate::util::copy_dir_all(&src, &target)?;
        append_copy(&src.join("copies.json"), medium, &target)?;
        append_copy(&target.join("copies.json"), medium, &target)?;
        Ok(target)
    }
}

pub fn validate_block_id(block_id: &str) -> Result<()> {
    let mut parts = block_id.split('-');
    let date = parts.next().unwrap_or("");
    let hex = parts.next().unwrap_or("");
    let ok = parts.next().is_none()
        && date.len() == 8
        && date.bytes().all(|b| b.is_ascii_digit())
        && hex.len() == 16
        && hex.bytes().all(|b| b.is_ascii_hexdigit());
    if ok {
        Ok(())
    } else {
        Err(Error::store(format!(
            "invalid block id {block_id:?}: expected YYYYMMDD- plus 16 hex chars"
        )))
    }
}

fn write_payload_sums(dest: &Path, payload: &Path) -> Result<()> {
    let mut files = vec![payload.to_path_buf()];
    if let Ok(rd) = std::fs::read_dir(dest) {
        for e in rd.flatten() {
            let p = e.path();
            if p.extension().and_then(|s| s.to_str()) == Some("par2") {
                files.push(p);
            }
        }
    }
    files.sort();
    files.dedup();
    hashing::write_sum_files(dest, &files)
}

fn append_copy(path: &Path, medium: &str, location: &Path) -> Result<()> {
    let mut copies = if path.exists() {
        manifest::read_json(path)?
    } else {
        serde_json::json!([])
    };
    if !copies.is_array() {
        copies = serde_json::json!([]);
    }
    if let Some(arr) = copies.as_array_mut() {
        arr.push(serde_json::json!({
            "medium": medium,
            "location": location.display().to_string(),
            "at": manifest::utcnow(),
        }));
    }
    manifest::write_json(path, &copies)
}
